File: encoder.py
from machine import Pin
import utime

class Encoder:
    def __init__(self, gpio_a, gpio_b, gpio_z, gpio_i):
        self.gpio_a = gpio_a
        self.gpio_b = gpio_b
        self.gpio_z = gpio_z
        self.gpio_i = gpio_i
        self.position = 0
        self.prev_position = 0
        self.prev_time = utime.ticks_us()
        self.encoder_pulses = 64
        self.timer_ticks_us = 125
        self.timer_interval_us = 10000
        self.init_gpio()
        self.init_interrupts()
    
    def init_gpio(self):
        self.pin_a = Pin(self.gpio_a, Pin.IN, Pin.PULL_UP)
        self.pin_b = Pin(self.gpio_b, Pin.IN, Pin.PULL_UP)
        self.pin_z = Pin(self.gpio_z, Pin.IN, Pin.PULL_UP)
        self.pin_i = Pin(self.gpio_i, Pin.IN, Pin.PULL_UP)
    
    def init_interrupts(self):
        self.pin_a.irq(handler=self.encoder_interrupt_A, trigger=(Pin.IRQ_RISING | Pin.IRQ_FALLING), priority=1)
        self.pin_b.irq(handler=self.encoder_interrupt_B, trigger=(Pin.IRQ_RISING | Pin.IRQ_FALLING), priority=1)
        self.pin_z.irq(handler=self.encoder_interrupt_Z, trigger=Pin.IRQ_RISING, priority=1)
        self.pin_i.irq(handler=self.encoder_interrupt_I, trigger=Pin.IRQ_RISING, priority=1)

    def get_rpm(self):
        current_time = utime.ticks_us()
        elapsed_time_us = current_time - self.prev_time
        position_change = self.position - self.prev_position
        rpm = (position_change * self.timer_ticks_us * self.encoder_pulses / elapsed_time_us) * 60.0
        self.prev_position = self.position
        self.prev_time = current_time
        return rpm
    
    def get_velocity(self):
        current_time = utime.ticks_us()
        elapsed_time_us = current_time - self.prev_time
        position_change = self.position - self.prev_position
        velocity = (position_change * self.timer_ticks_us * self.encoder_pulses / elapsed_time_us) * (self.timer_ticks_us / 1000000.0)
        self.prev_position = self.position
        self.prev_time = current_time
        return velocity
    
    def get_acceleration(self):
        current_time = utime.ticks_us()
        elapsed_time_us = current_time - self.prev_time
        position_change = self.position - self.prev_position
        acceleration = (position_change * self.timer_ticks_us * self.encoder_pulses / elapsed_time_us) * (self.timer_ticks_us / 1000000.0) / (elapsed_time_us / 1000000.0)
        self.prev_position = self.position
        self.prev_time = current_time
        return acceleration
    
    def get_direction(self):
        position_change = self.position - self.prev_position
        direction = '+' if position_change > 0 else '-'
        self.prev_position = self.position
        return direction
    
    def reset_position(self):
        self.position = 0
        self.prev_position = 0

File: encoder.cpp
#include "encoder.h"

Encoder::Encoder(uint gpio_a, uint gpio_b, uint gpio_z, uint gpio_i) {
    m_gpio_a = gpio_a;
    m_gpio_b = gpio_b;
    m_gpio_z = gpio_z;
    m_gpio_i = gpio_i;
    gpio_init(m_gpio_a);
    gpio_set_dir(m_gpio_a, GPIO_IN);
    gpio_pull_up(m_gpio_a);
    gpio_init(m_gpio_b);
    gpio_set_dir(m_gpio_b, GPIO_IN);
    gpio_pull_up(m_gpio_b);
    gpio_init(m_gpio_z);
    gpio_set_dir(m_gpio_z, GPIO_IN);
    gpio_pull_up(m_gpio_z);
    gpio_init(m_gpio_i);
    gpio_set_dir(m_gpio_i, GPIO_IN);
    gpio_pull_up(m_gpio_i);
    gpio_set_irq_enabled_with_callback(m_gpio_a, GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL, true, &Encoder::encoder_interrupt_A);
    gpio_set_irq_enabled_with_callback(m_gpio_b, GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL, true, &Encoder::encoder_interrupt_B);
    gpio_set_irq_enabled_with_callback(m_gpio_z, GPIO_IRQ_EDGE_RISE, true, &Encoder::encoder_interrupt_Z);
    gpio_set_irq_enabled_with_callback(m_gpio_i, GPIO_IRQ_EDGE_RISE, true, &Encoder::encoder_interrupt_I);
    m_prev_time = timer_get_us();
}

double Encoder::get_rpm() {
    uint64_t current_time = timer_get_us();
    uint64_t elapsed_time_us = current_time - m_prev_time;
    int position_change = m_position - m_prev_position;
    double rpm = (double)position_change * (double)m_timer_ticks_us * (double)m_encoder_pulses / (double)elapsed_time_us * 60.0;
    m_prev_position = m_position;
    m_prev_time = current_time;
    return rpm;
}

double Encoder::get_velocity() {
    uint64_t current_time = timer_get_us();
    uint64_t elapsed_time_us = current_time - m_prev_time;
    int position_change = m_position - m_prev_position;
    double velocity = (double)position_change * (double)m_timer_ticks_us * (double)m_encoder_pulses / (double)elapsed_time_us * (double)m_timer_ticks_us / 1000000.0;
    m_position = m_position;
m_prev_time = current_time;
return velocity;
}

double Encoder::get_acceleration() {
uint64_t current_time = timer_get_us();
uint64_t elapsed_time_us = current_time - m_prev_time;
int position_change = m_position - m_prev_position;
double acceleration = (double)position_change * (double)m_timer_ticks_us * (double)m_encoder_pulses / (double)elapsed_time_us * (double)m_timer_ticks_us / 1000000.0 / (double)elapsed_time_us * 1000000.0;
m_prev_position = m_position;
m_prev_time = current_time;
return acceleration;
}

char Encoder::get_direction() {
int position_change = m_position - m_prev_position;
char direction = (position_change > 0) ? '+' : '-';
m_prev_position = m_position;
return direction;
}

void Encoder::reset_position() {
m_position = 0;
m_prev_position = 0;
}

void Encoder::encoder_interrupt_A() {
Encoder encoder = reinterpret_cast<Encoder>(gpio_get_irq_user_data(Encoder::m_gpio_a));
if (gpio_get(encoder->m_gpio_a) == gpio_get(encoder->m_gpio_b)) {
encoder->m_position++;
} else {
encoder->m_position--;
}
}

void Encoder::encoder_interrupt_B() {
Encoder encoder = reinterpret_cast<Encoder>(gpio_get_irq_user_data(Encoder::m_gpio_b));
if (gpio_get(encoder->m_gpio_a) == gpio_get(encoder->m_gpio_b)) {
encoder->m_position--;
} else {
encoder->m_position++;
}
}

void Encoder::encoder_interrupt_Z() {
Encoder encoder = reinterpret_cast<Encoder>(gpio_get_irq_user_data(Encoder::m_gpio_z));
encoder->reset_position();
}

void Encoder::encoder_interrupt_I() {
Encoder encoder = reinterpret_cast<Encoder>(gpio_get_irq_user_data(Encoder::m_gpio_i));
// Do something when interrupt pin I is triggered
printf("Interrupt pin I triggered!\n");
}

void Encoder::timer_interrupt() {
uint64_t current_time = timer_get_us();
double rpm = get_rpm();
double velocity = get_velocity();
double acceleration = get_acceleration();
char direction = get_direction();
printf("RPM: %.2f, Velocity: %.2f %c/s, Acceleration: %.2f %c/s^2, Direction: %c\n", rpm, velocity, direction, acceleration, direction, direction);
}


